### PR TITLE
[Magiclysm]: Holiday Monsters Part II

### PR DESCRIPTION
**Pull request by Theawesomeboophis**
#### Summary

SUMMARY: Mods "Adds Workshop Elves, Yule Cats, and Leprechauns to Magiclysm"

#### Purpose of change
Adds The Above creatures to Magiclysm as part of #45672 

#### Describe the solution
Adds Workshop Elves( Regular Elves coming later, thus the distinction), Yule Cats, and Leprechauns.

#### Describe alternatives you've considered
Not Doing This
Making Workshop Elves Group and moving them to new worldgen location (Might do in future PR)

#### Testing
All errors found while testing fixed, creatures spawn and look right with correct colors.
#### Additional context
Part II of #45672, Part III will be adding MSC monsters described in the issue.

### CODE REVIEW

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
+[
+  {
+    "id": "mon_yulecat",
+    "type": "MONSTER",
+    "name": "Yule Cat",
+    "description": " Magical beast from Icelandic folklore, known for their catlike appearance and their hobby of eating Workshop Elves.",
```
Comment (BrettDong) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:6]: Is the starting space here intentional or a typo?

Comment (Theawesomeboophis) [reply data/mods/Magiclysm/monsters/holiday_magiclysm.json:6]: Typo.

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    "melee_cut": 7,
+    "dodge": 6,
+    "harvest": "mammal_fur",
+    "special_attacks": [ [ "EAT_FOOD", 71 ] ],
+    "upgrades": { "age_grow": 234, "into": "mon_yulecat" }
+  },
+  {
+    "id": "mon_workshop_elf",
+    "type": "MONSTER",
+    "name": "Workshop Elf",
```
Comment (BrettDong) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:57]: Needs to manually write down its plural form
```suggestion
    "name": { "str": "Workshop Elf", "str_pl": "Workshop Elves" },
```

**Comment by Theawesomeboophis** [comment]
Might add pots of gold to Leprechauns, but as that is a modern leprechaun and not a mythological Leprechaun, might leave leprechauns without them (Original Leprechauns were a type of fairy in Irish Mythology, the modern depictions of leprechauns are from modern folklore and popular culture).

**Comment by Theawesomeboophis** [comment]
Testing was successful, after I fixed the separator and added default faction.

**Review by actual-nh — COMMENTED** [review]
Interesting; not sure if I would include the elves or leprechauns except if Crazycataclysm was also on...

**Comment by Theawesomeboophis** [comment]
I tried adding Path_Avoid_Danger 2 to the dinosaurs (From Obscure Dinos
Issue) and It gave an error saying it wasn't a real flag, I might have
typed it wrong last time but going to check flag files just in case.

**Comment by KorGgenT** [comment]
i'm going to say "no" on workshop elves and the things relating to them. If elves are added to magiclysm, i would want them to be the LOTR/D&D style elves.

**Comment by actual-nh** [comment]
> I tried adding Path_Avoid_Danger 2 to the dinosaurs (From Obscure Dinos Issue) and It gave an error saying it wasn't a real flag, I might have typed it wrong last time but going to check flag files just in case.

PATH_AVOID_DANGER_2 is one that is easy to mistype (e.g.,  by leaving out the last underline). There are quite a lot of monsters with it.

**Comment by Theawesomeboophis** [comment]
> i'm going to say "no" on workshop elves and the things relating to them. If elves are added to magiclysm, i would want them to be the LOTR/D&D style elves.

I will be adding those elves in the future as well.

**Comment by KorGgenT** [comment]
@Theawesomeboophis i _really_ appreciate all the work you're doing recently, but is there any way you can combine some of those commits? each single individual commit is an email i receive. i've gotten 200 emails in the past 14 days, which overwhelms me and i can't figure out which ones are important. i would be happy to help you with this on the discord https://discord.gg/cdda

**Comment by Theawesomeboophis** [comment]
> @Theawesomeboophis i _really_ appreciate all the work you're doing recently, but is there any way you can combine some of those commits? each single individual commit is an email i receive. i've gotten 200 emails in the past 14 days, which overwhelms me and i can't figure out which ones are important. i would be happy to help you with this on the discord https://discord.gg/cdda

How do I combine commits.

**Comment by KorGgenT** [comment]
so much of our development is done on the discord, but if you don't want to use it you don't want to use it.

There's two things at play here: just now you did a bunch of individual commits from a review. you can actually batch those and commit them all at once at the end, which is only 1 email, instead of 5. 

The other thing is using a local copy of the repo and a git client of some kind (i use github desktop, highly recommend) and work on there. then commit when you have accumulated the changes you want that all logically work together (and push). in that case, i only get 1 email each time someone pushes from their local repo.

thank you again.

**Comment by actual-nh** [comment]
> @Theawesomeboophis i _really_ appreciate all the work you're doing recently, but is there any way you can combine some of those commits? each single individual commit is an email i receive.

Oh, dear. I am not sure how to do that with code review commits (as in accepting suggestions).

**Comment by Theawesomeboophis** [comment]
> > @Theawesomeboophis i _really_ appreciate all the work you're doing recently, but is there any way you can combine some of those commits? each single individual commit is an email i receive.
> 
> Oh, dear. I am not sure how to do that with code review commits (as in accepting suggestions).

Actually, I can add them to a batch which I believe makes them group together and then just merge them all at once.

**Comment by Theawesomeboophis** [comment]
> so much of our development is done on the discord, but if you don't want to use it you don't want to use it.
> 
> There's two things at play here: just now you did a bunch of individual commits from a review. you can actually batch those and commit them all at once at the end, which is only 1 email, instead of 5.
> 
> The other thing is using a local copy of the repo and a git client of some kind (i use github desktop, highly recommend) and work on there. then commit when you have accumulated the changes you want that all logically work together (and push). in that case, i only get 1 email each time someone pushes from their local repo.
> 
> thank you again.

Orcs and Ogres is complete now and works properly btw, you told me to notify you when it was done in #46187, but I don't think the email stood out amongst the others. (Being someone with 800 emails a day, I have the same problem).

**Comment by actual-nh** [comment]
> > > @Theawesomeboophis i _really_ appreciate all the work you're doing recently, but is there any way you can combine some of those commits? each single individual commit is an email i receive.
> > 
> > 
> > Oh, dear. I am not sure how to do that with code review commits (as in accepting suggestions).
> 
> Actually, I can add them to a batch which I believe makes them group together and then just merge them all at once.

I was just about to email you with a [link](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request) on that.

**Comment by Theawesomeboophis** [comment]
Merging Orcs And Ogres seems to have created some problems in monster factions, How do I fix that, by merging master into branch or is there another way?

**Comment by actual-nh** [comment]
> so much of our development is done on the discord, but if you don't want to use it you don't want to use it.

Umm... I don't have any of the accounts that are usable to get through the AltDentifier screening (since Captcha isn't one of them).

**Comment by Theawesomeboophis** [comment]
Accidentally merged master branch of my fork, hang on.

**Comment by KorGgenT** [comment]
> > so much of our development is done on the discord, but if you don't want to use it you don't want to use it.
> 
> Umm... I don't have any of the accounts that are usable to get through the AltDentifier screening (since Captcha isn't one of them).

i gotchu.

**Comment by Theawesomeboophis** [comment]
Merge conflicts resolved, merged Upstream master instead of Boophis master this time.

**Comment by Theawesomeboophis** [comment]
Going to check for more old orc remnants that may be in the files.

**Comment by Theawesomeboophis** [comment]
It seems that I have broken Obscure Dinosaurs as well.

**Comment by Theawesomeboophis** [comment]
This is all sorts of broken, how do I fix Dinomod removals?

**Comment by Theawesomeboophis** [comment]
Drafted as it seems a whole bunch of files were altered unintentionally.

**Comment by KorGgenT** [comment]
well you could `git rebase -i upstream/master` and remove all the unecessary commits

**Comment by Theawesomeboophis** [comment]
> well you could `git rebase -i upstream/master` and remove all the unecessary commits

Thanks, I was just replacing modified files with base counterparts, but your method seems faster.

**Comment by Theawesomeboophis** [comment]
There we go, goodbye dupes and legacy code.

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    "default_faction": "leprechaun",
+    "bodytype": "human",
+    "volume": "145 ml",
+    "weight": "168 kg",
+    "hp": 125,
+    "speed": 124,
+    "material": [ "flesh" ],
+    "symbol": "L",
+    "color": "green",
+    "aggression": 100,
```
[the remark this answers is not in the document]
Comment (Theawesomeboophis) [reply data/mods/Magiclysm/monsters/holiday_magiclysm.json:100]: Not sure how to implement that, there may be a flag for it however.

**Comment by BrettDong** [comment]
Commits here in this PR looks messy. Could you put your changes in one or a few commits and force-push to this PR?

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    "extend": { "flags": [ "NO_BREED" ], "fear_triggers": [ "PLAYER_CLOSE", "HURT" ] },
+    "delete": { "anger_triggers": "PLAYER_NEAR_BABY" },
+    "upgrades": { "age_grow": 234, "into": "mon_yulecat" }
+  },
+  {
+    "id": "mon_workshop_elf",
+    "type": "MONSTER",
+    "name": { "str": "Workshop Elf", "str_pl": "Workshop Elves" },
+    "description": "Magical humanoids who are known for working in workshops, wearing green, and being proficient at making toys.",
+    "default_faction": "elves",
```
Comment (BrettDong) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:64]: Game does not load and complains:
```
 DEBUG    : Monster type 'mon_workshop_elf' has invalid default_faction: 'elves'. Add this faction to json as MONSTER_FACTION type.

 FUNCTION : void MonsterGenerator::finalize_mtypes()
 FILE     : src/monstergenerator.cpp
 LINE     : 373
```

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    "harvest": "human",
+    "death_function": [ "NORMAL" ],
+    "flags": [ "SEES", "HEARS", "SMELLS", "PATH_AVOID_DANGER_2", "WARM" ]
+  },
+  {
+    "id": "mon_leprechaun",
+    "type": "MONSTER",
+    "name": "Leprechaun",
+    "description": "Mischievous shoemakers who wear green and love pulling pranks on mortals.",
+    "default_faction": "leprechaun",
```
Comment (BrettDong) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:91]: Another error here:
```
 DEBUG    : Monster type 'mon_leprechaun' has invalid default_faction: 'leprechaun'. Add this faction to json as MONSTER_FACTION type.

 FUNCTION : void MonsterGenerator::finalize_mtypes()
 FILE     : src/monstergenerator.cpp
 LINE     : 373
```

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json (hunk 4 above)
Comment (Theawesomeboophis) [reply data/mods/Magiclysm/monsters/holiday_magiclysm.json:64]: I did actually create monster factions, they were accidentally deleted while fixing merge conflicts, I'll add them back, thanks for notifying me.

**Comment by KorGgenT** [comment]
i agree with @BrettDong here on this one, please rebase by using `git rebase -i upstream/master` in CLI on your machine, and squash your commits down. the orc PR was ok but this one is undoing and redoing way too much to let slide in my opinion.

**Comment by Theawesomeboophis** [comment]
Can I rebase using desktop, and if I can, how do I do so?

**Comment by BrettDong** [comment]
Assume `master` branch is up-to-date, in command line do
```sh
git checkout master
git checkout -b tmp
git merge --squash Holiday-Monsters-Part-II
git commit -m 'Magiclysm: Holiday Monsters Part II'
git checkout Holiday-Monsters-Part-II
git reset --hard master
git cherry-pick tmp
git push --force
git branch -D tmp
```

**Comment by KorGgenT** [comment]
you won't be able to squash commits in desktop, you will need to open up your command line, which should be accessible in desktop here 
image

**Comment by Theawesomeboophis** [comment]
What do I type into command prompt? I attempted to type in @BrettDong's lines but it said git is an invalid command.

**Comment by BrettDong** [comment]
Just helped you squash the messed commit history. Now you can delete the local branch on your machine and re-pull from GitHub to get in sync.

**Comment by Theawesomeboophis** [comment]
How did I undo it? I am really bad at this.

**Comment by Theawesomeboophis** [comment]
Oh no.

**Comment by Theawesomeboophis** [comment]
How do I fix this?

**Comment by BrettDong** [comment]
After I clean up this branch again, delete your copy of this branch on your machine, and pull this branch from GitHub as if it is brand new

**Comment by Theawesomeboophis** [comment]
> After I clean up this branch again, delete your copy of this branch on your machine, and pull this branch from GitHub as if it is brand new

Ok, sorry about the inconvenience, I don't know what I'm doing sometimes.

**Comment by BrettDong** [comment]
Why are you still pushing irrelevant commits?

**Comment by Theawesomeboophis** [comment]
> Why are you still pushing irrelevant commits?

I'm reverting the files back to their base counterparts.

**Comment by BrettDong** [comment]
In this case, please do _reset_ instead of _revert_

**Comment by Theawesomeboophis** [comment]
> In this case, please do _reset_ instead of _revert_

I'm not really reverting, I'm just copying and pasting the contents of the changed files from base to undo changes unintentionally committed to them.

**Comment by BrettDong** [comment]
What needs to be done here is not just reverting unintentionally committed changes, but cleaning up commit history. A reset does the trick.

**Comment by Theawesomeboophis** [comment]
Accidentally closed, trying to figure out mobile.

**Comment by Maleclypse** [comment]
It looks like you are ready for final review on this?

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 6]
```diff
@@ -0,0 +1,110 @@
+[
+  {
+    "id": "mon_yulecat",
+    "type": "MONSTER",
+    "name": "Yule Cat",
+    "description": "Magical beast from Icelandic folklore, known for their catlike appearance and their hobby of eating Workshop Elves.",
+    "bodytype": "lion",
+    "default_faction": "magical_beast",
+    "species": [ "MAGICAL_BEAST" ],
+    "volume": "200000 ml",
```
Comment (Maleclypse) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:10]:
```suggestion
    "volume": "200 L",
```

File: data/mods/Magiclysm/monstergroups.json [hunk 7]
```diff
@@ -16,6 +16,9 @@
       { "monster": "mon_troll", "freq": 3, "cost_multiplier": 10, "conditions": [ "DAY" ] },
       { "monster": "mon_claygolem", "freq": 8, "cost_multiplier": 3 },
       { "monster": "mon_stonegolem", "freq": 4, "cost_multiplier": 5 },
+      { "monster": "mon_yulecat", "freq": 2, "cost_multiplier": 7 },
```
Comment (Maleclypse) [inline data/mods/Magiclysm/monstergroups.json:19]: I would delay these from appearing until late fall or winter.  Let me know if you need help finding the json to do that.  With as fast and tough as they are can a day one character escape from them?

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    "default_faction": "leprechaun",
+    "bodytype": "human",
+    "volume": "145 ml",
+    "weight": "168 kg",
+    "hp": 125,
+    "speed": 124,
+    "material": [ "flesh" ],
+    "symbol": "L",
+    "color": "green",
+    "aggression": 100,
```
Comment (Maleclypse) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:100]: They are 100% hostile?

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    "morale": 100,
+    "melee_skill": 6,
+    "melee_dice": 7,
+    "melee_dice_sides": 7,
+    "dodge": 6,
+    "harvest": "human",
```
Comment (Maleclypse) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:106]:
```suggestion
    "harvest": "demihuman",
```

File: data/mods/Magiclysm/monsters/holiday_magiclysm.json [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    "aggression": 12,
+    "morale": 100,
+    "melee_skill": 4,
+    "melee_dice": 5,
+    "melee_dice_sides": 6,
+    "dodge": 3,
+    "armor_bash": 6,
+    "armor_cut": 2,
+    "armor_bullet": 3,
+    "harvest": "human",
```
Comment (Maleclypse) [inline data/mods/Magiclysm/monsters/holiday_magiclysm.json:82]:
```suggestion
    "harvest": "demihuman",
```

**Review by Maleclypse — COMMENTED** [review]
Had a couple questions.  Let me know if you need any help.  Feel free to ping me here or discord.

File: data/mods/Magiclysm/monstergroups.json (hunk 7 above)
Comment (Theawesomeboophis) [reply data/mods/Magiclysm/monstergroups.json:19]: Will adding conditions: winter work for this. ( Like how troll has conditions: day)?

**Comment by Theawesomeboophis** [comment]
The next thing I will be working on after Part III will be expanding elves
( Possibly gonna add light and dark elves dnd style with variants,
considering a Elf King who will act as a boss-type enemy).

**Pull request merged**